### PR TITLE
Find Camera Viewport

### PR DESCRIPTION
**Pull request by SirePi**
Implementing issue #681.
In addition, added a shorthand `FocusPos` property to have quick access to the current focused position of the Camera.

### CODE REVIEW

**Comment by SirePi** [comment]
Might have an issue.. at the moment I am calculating the corners with a different imageSize starting from the top left = (0,0).. should I center the imageSize on the screen first instead?

**Comment by ilexp** [comment]
> Might have an issue.. at the moment I am calculating the corners with a different imageSize starting from the top left = (0,0).. should I center the imageSize on the screen first instead?

Not 100% sure that answers the question, but the rect of imageSize should be centered around the camera position, so that the rects center is the focus point of the camera.

File: Source/Core/Duality/Components/Camera.cs [hunk 1]
```diff
@@ -333,6 +438,7 @@ private void SetupDrawDevice()
 			// It will never be used by the Cameras transform methods.
 			this.drawDevice = new DrawDevice();
 		}
+
```
Comment (ilexp) [inline Source/Core/Duality/Components/Camera.cs]: Extra newline - please remove.

File: Source/Core/Duality/Components/Camera.cs [hunk 2]
```diff
@@ -324,6 +335,100 @@ public bool IsSphereInView(Vector3 worldPos, float radius = 1.0f)
...
+		{
+			return this.GetWorldViewportBounds(z, DualityApp.TargetViewSize);
+		}
+
+		/// <summary>
+		/// Return an axis-aligned Rect that contains the current viewport with the desired size.
+		/// </summary>
+		/// <param name="imageSize">Target size of the rendered image before adjusting it to fit the specified viewport.</param>
+		/// <returns>An axis-aligned Rect that fits the viewport. Could be bigger if the Camera is not rotated in 90° increments.</returns>
+		public Rect GetWorldViewportBounds(Vector2 imageSize)
```
[the remark this answers is not in the document]
Comment (SirePi) [reply Source/Core/Duality/Components/Camera.cs]: ok

File: Source/Core/Duality/Components/Camera.cs [hunk 3]
```diff
@@ -324,6 +335,100 @@ public bool IsSphereInView(Vector3 worldPos, float radius = 1.0f)
...
+		{
+			this.GetWorldViewportCorners(corners, z, DualityApp.TargetViewSize);
+		}
+
+		/// <summary>
+		/// Fills an array with vectors that correspond to the 4 corners of the current viewport with the desired size.
+		/// </summary>
+		/// <param name="corners">An array that will contain the corners of the viewport, in counter-clockwise order, starting from the top-left. The array must fit at least 4 elements.</param>
+		/// <param name="imageSize">Target size of the rendered image before adjusting it to fit the specified viewport.</param>
+		public void GetWorldViewportCorners(Vector2[] corners, Vector2 imageSize)
```
[the remark this answers is not in the document]
Comment (SirePi) [reply Source/Core/Duality/Components/Camera.cs]: ok

File: Source/Core/Duality/Components/Camera.cs [hunk 4]
```diff
@@ -324,6 +335,100 @@ public bool IsSphereInView(Vector3 worldPos, float radius = 1.0f)
...
+		/// <summary>
+		/// Return an axis-aligned Rect that contains the current viewport with the desired size.
+		/// </summary>
+		/// <param name="z">Z world-coordinate used to determine the extents of the viewport.</param>
+		/// <returns>An axis-aligned Rect that fits the viewport. Could be bigger if the Camera is not rotated in 90° increments.</returns>
+		public Rect GetWorldViewportBounds(float z)
+		{
+			return this.GetWorldViewportBounds(z, DualityApp.TargetViewSize);
+		}
+
```
[the remark this answers is not in the document]
Comment (SirePi) [reply Source/Core/Duality/Components/Camera.cs]: ok

File: Source/Core/Duality/Components/Camera.cs [hunk 5]
```diff
@@ -324,6 +335,100 @@ public bool IsSphereInView(Vector3 worldPos, float radius = 1.0f)
...
+			float minY = float.PositiveInfinity;
+			float maxX = float.NegativeInfinity;
+			float maxY = float.NegativeInfinity;
+
+			for (int i = 0; i < 4; i++)
+			{
+				minX = MathF.Min(minX, this.cameraBounds[i].X);
+				minY = MathF.Min(minY, this.cameraBounds[i].Y);
+				maxX = MathF.Max(maxX, this.cameraBounds[i].X);
+				maxY = MathF.Max(maxY, this.cameraBounds[i].Y);
```
[the remark this answers is not in the document]
Comment (SirePi) [reply Source/Core/Duality/Components/Camera.cs]: oook

File: Source/Core/Duality/Components/Camera.cs (hunk 1 above)
Comment (SirePi) [reply Source/Core/Duality/Components/Camera.cs]: ooooooooook

File: Source/Core/Duality/Components/Camera.cs [hunk 6]
```diff
@@ -70,6 +71,16 @@ public float FocusDist
 			get { return this.focusDist; }
 			set { this.focusDist = MathF.Max(value, 0.01f); }
 		}
+
```
[the remark this answers is not in the document]
Comment (SirePi) [reply Source/Core/Duality/Components/Camera.cs]: ok

**Pull request merged**